src/qpx_scan.cpp

```diff
@@ -0,0 +1,470 @@
+
+  //	pluginlib = dlopen( pname, RTLD_NOW | RTLD_GLOBAL);
+  pluginlib = dlopen(pname, RTLD_LAZY | RTLD_GLOBAL);
+  if (!pluginlib) {
+    printf("0 dlopen err: %s\n", dlerror());
+    //		printf("can't open library!\n");
+    goto plugin_attach_liberr;
+  }
+//	if (!dev->silent) printf("pluginlib = %p\n", pluginlib);
+#ifndef _WIN32
+  if (dlerror()) {
+    printf("1 dlopen err: %s\n", dlerror());
+    goto plugin_attach_err;
+  }
+#endif
+  if (!dev->silent)
+    printf("plugin lib opened: %s\n", pname);
+  plugin_create =
+      (scan_plugin * (*)(drive_info *)) dlsym(pluginlib, "plugin_create");
+#ifndef _WIN32
+  if (dlerror()) {
+#else
+  if (!plugin_create) {
+#endif
+    printf("error searching symbol \"plugin_create\" : %s\n", dlerror());
+    goto plugin_attach_err;
+  }
+  //	printf("symbol \"plugin_create\" found!\n");
+  //*(void **) (&plugin_destroy) = dlsym(pluginlib, "plugin_destroy");
+  plugin_destroy = (void (*)(scan_plugin *))dlsym(pluginlib, "plugin_destroy");
+#ifndef _WIN32
+  if (dlerror()) {
+#else
+  if (!plugin_destroy) {
+#endif
+    printf("error searching symbol \"plugin_destroy\" : %s\n", dlerror());
+    goto plugin_attach_err;
+  }
+  //	printf("symbol \"plugin_destroy\" found!\n");
+  /*
+      slist = (drivedesc*) dlsym(pluginlib, "devlist");
+      e = dlerror();
+      if (e){
+              printf("%s\nerror searching symbol \"devlist\"\n",e);
+              goto plugin_attach_err;
+      }
+  */
+  //	printf("creating plugin\n");
+  plugin = plugin_create(dev);
+  //	printf("plugin info\n");
+  if (!silent)
+    printf("Found plugin: %s (%s)\n", plugin->name(), plugin->desc());
+
+  if (plugin->blklist) {
+    devlist = plugin->blklist;
+    if (!dev->silent) {
+      printf("Devices in blacklist:\n");
+      for (int d = 0; devlist[d].ven_ID > 0; d++) {
+        printf("  %s %s*\n", devlist[d].ven, devlist[d].dev);
+      }
+    }
+
+    for (int d = 0; !blacklisted && devlist[d].ven_ID > 0; d++) {
+      if (!strncmp(dev->ven, devlist[d].ven, strlen(devlist[d].ven)) &&
+          !strncmp(dev->dev, devlist[d].dev, strlen(devlist[d].dev)))
+        blacklisted = 1;
+    }
+
+    devlist = NULL;
+    if (blacklisted) {
+      printf("Plugin %s: device '%s' '%s' blacklisted! Detaching plugin...\n",
+             plugin->name(), dev->ven, dev->dev);
+      attached = 1;
+      plugin_detach();
+      return 1;
+    }
+  }
+
+  if (!probe_enable && plugin->devlist) {
+    devlist = plugin->devlist;
+    dev->chk_features = 0;
+
+    if (!dev->silent) {
+      printf("Devices supported by this plugin:\n");
+      for (int d = 0; devlist[d].ven_ID > 0 && devlist[d].dev_ID > 0; d++) {
+        printf("  %s %s\n", devlist[d].ven, devlist[d].dev);
+      }
+    }
+
+    for (int d = 0; !listed && devlist[d].ven_ID > 0 && devlist[d].dev_ID > 0;
+         d++) {
+      if (!strncmp(dev->ven, devlist[d].ven, strlen(devlist[d].ven)) &&
+          !strncmp(dev->dev, devlist[d].dev, strlen(devlist[d].dev))) {
```

Comment:
Codacy found an issue: [Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected) (CWE-126).  ](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)